opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -338,39 +353,97 @@ private ExprValue createOpenSearchDateType(Content value, ExprType type) {
+  private ExprValue parseArray(
+      Content content,
+      String prefix,
+      ExprType type,
+      boolean supportArrays
+  ) {
     List<ExprValue> result = new ArrayList<>();
-    // ExprCoreType.ARRAY does not indicate inner elements type.
-    if (Iterators.size(content.array()) == 1 && content.objectValue() instanceof JsonNode) {
-      result.add(parse(content, prefix, Optional.of(STRUCT)));
+
+    // ARRAY is mapped to nested but can take the json structure of an Object.
+    if (content.objectValue() instanceof ObjectNode) {
+      result.add(parseStruct(content, prefix, supportArrays));
+      // non-object type arrays are only supported when parsing inner_hits of OS response.
+    } else if (
+        !(type instanceof OpenSearchDataType
+            && ((OpenSearchDataType) type).getExprType().equals(ARRAY))
+        && !supportArrays) {
+      return parseInnerArrayValue(content.array().next(), prefix, type, supportArrays);
     } else {
       content.array().forEachRemaining(v -> {
-        // ExprCoreType.ARRAY does not indicate inner elements type. OpenSearch nested will be an
-        // array of structs, otherwise parseArray currently only supports array of strings.
-        if (v.isString()) {
-          result.add(parse(v, prefix, Optional.of(OpenSearchDataType.of(STRING))));
-        } else {
-          result.add(parse(v, prefix, Optional.of(STRUCT)));
-        }
+        result.add(parseInnerArrayValue(v, prefix, type, supportArrays));
       });
     }
     return new ExprCollectionValue(result);
   }
 
+  /**
+   * Parse inner array value. Can be object type and recurse continues.
+   * @param content Array index being parsed.
+   * @param prefix Prefix for value.
+   * @param type Type of inner array value.
+   * @param supportArrays Parsing the whole array if array is type nested.
+   * @return Inner array value.
+   */
+  private ExprValue parseInnerArrayValue(
+      Content content,
+      String prefix,
+      ExprType type,
+      boolean supportArrays
+  ) {
+    if (type instanceof OpenSearchIpType
+        || type instanceof OpenSearchBinaryType
+        || type instanceof OpenSearchDateType
+        || type instanceof OpenSearchGeoPointType) {
+      return parse(content, prefix, Optional.of(type), supportArrays);
+    } else if (content.isString()) {
+      return parse(content, prefix, Optional.of(OpenSearchDataType.of(STRING)), supportArrays);
+    } else if (content.isLong()) {
+      return parse(content, prefix, Optional.of(OpenSearchDataType.of(LONG)), supportArrays);
+    } else if (content.isFloat()) {
+      return parse(content, prefix, Optional.of(OpenSearchDataType.of(FLOAT)), supportArrays);
+    } else if (content.isDouble()) {
+      return parse(content, prefix, Optional.of(OpenSearchDataType.of(DOUBLE)), supportArrays);
+    } else if (content.isNumber()) {
+      return parse(content, prefix, Optional.of(OpenSearchDataType.of(INTEGER)), supportArrays);
```

Comment:
If you are using something with type of `SHORT` or `BYTE` does it get cast/treated as an `INTEGER` before this point or does it need to be supported?

Reply:
Treated as an integer. No need to cast or treat differently.